File: src/leetCodeProblems/DFSUsingRecursion/BinarySearchTree/BSTRangeSum938.go
package main

/*
- LeetCode - https://leetcode.com/problems/range-sum-of-bst/
- Time - O(logn)
- Space - O(logn) (considering the stack used in recursion)
*/

func rangeSumBSTUtil(root *TreeNode, low, high int, sum *int) {
	if root == nil {
		return
	}

	if low <= root.Val && root.Val <= high {
		*sum += root.Val
	}

	rangeSumBSTUtil(root.Left, low, high, sum)
	rangeSumBSTUtil(root.Right, low, high, sum)
}

func rangeSumBST(root *TreeNode, low int, high int) int {
	sum := 0
	rangeSumBSTUtil(root, low, high, &sum)
	return sum
}

// func main() {
// 	root := new(TreeNode)
// 	root.Val = 4
// 	root.Left = new(TreeNode)
// 	root.Left.Val = 2
// 	root.Right = new(TreeNode)
// 	root.Right.Val = 7
// 	root.Left.Left = new(TreeNode)
// 	root.Left.Left.Val = 1
// 	root.Left.Right = new(TreeNode)
// 	root.Left.Right.Val = 3

// 	//root = searchBST(root, 2)
// 	root = searchBST(root, 5)

// 	log.Println("root", root)
// }

File: src/leetCodeProblems/DFSUsingRecursion/BinarySearchTree/BSTRangeSum938.java
package leetCodeProblems.BinarySearchTree;

/**
 * https://leetcode.com/problems/range-sum-of-bst/
 */
public class BSTRangeSum938 {

    static class TreeNode {

        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }

    public int rangeSumBST(TreeNode root, int low, int high) {

        if (root == null) {
            return 0;
        }

        int sum = rangeSumBST(root.left, low, high);

        if (low <= root.val && root.val <= high) {
            sum += root.val;
        }

        sum += rangeSumBST(root.right, low, high);

        return sum;
    }

    public static void main(String[] args) {

        BSTRangeSum938 ob = new BSTRangeSum938();

        /*TreeNode root = new TreeNode(10);
        root.left = new TreeNode(5);
        root.right = new TreeNode(15);
        root.left.left = new TreeNode(3);
        root.left.right = new TreeNode(7);
        root.right.right = new TreeNode(18);*/

        //int sum = ob.rangeSumBST(root, 7, 15);

        TreeNode root = new TreeNode(10);
        root.left = new TreeNode(5);
        root.right = new TreeNode(15);
        root.left.left = new TreeNode(3);
        root.left.left.left = new TreeNode(1);
        root.left.right = new TreeNode(7);
        root.left.right.left = new TreeNode(6);
        root.right.right = new TreeNode(18);

        System.out.println(ob.rangeSumBST(root, 6, 10));
    }
}
